### ISSUE TITLE
Remove rapydo core

### ISSUE THREAD

**Issue by mdantonio** [rapydo/issues#79]
I'm unable to find an issue about this topic, so i'm opening a new one

We already discussed about the removal of rapydo core

We created rapydo core to collect everything not strictly depending from a single repo (backend/frontend). For example nginx is related to both of them

Refactor after refactor, now we inherit only for the content of conf folder:

```
backend.yml
frontend.yml
logging.yml
elasticsearch
kibana
logstash 
nginx
```

To keep updated such files we have to:
- manage upstream remote on git
- periodically pull from upstream

The period pull often brings to conflicts and it downloads unwanted folders (e.g. all vanilla projects used for tests)

This management is quite cumbersome, in order to simply keep some static piece of code...

If we move such compose yamls and such configuration folders somewhere, we could remove rapydo core (or better, rename it as rapydo-tests, as in fact it is already)

Options are:

- move everything into a common repository, like utils (where project_defaults already is)
- move on the more related repo, e.g. backend.yaml into backend, logging.yaml into utils (more or less how was before the refactors)
- create a new rapydo-confs repo downloaded as new submodules and also move there project_defaults

The more complicated thing is to decide with option implement, then this fix should be quite quick (we only have to modify paths in compose files)

**Comment by pdonorio** [comment]
Also [templating](https://github.com/rapydo/issues/issues/22#issuecomment-341001785) may help in someway here.

**Comment by mdantonio** [comment]
The problem with templating is not the update? We need to update such files in an easy way

**Comment by pdonorio** [comment]
Yes, and that should be investigated.

**Comment by pdonorio** [comment]
Anyway the sweetest option here is:

> create a new rapydo-confs repo downloaded as new submodules and also move there project_defaults

**Comment by pdonorio** [comment]
This depends on #22

**Comment by mdantonio** [comment]
restclient mounts this, to be fixed

- ../submodules/utilities/utilities/projects_defaults.yaml:/code/base.yml

The restclient images already obtains this file from:

dfile="/usr/local/lib/python3.6/site-packages/utilities/projects_defaults.yaml"

and not from:

# dfile="/code/base.yml"

So @pdonorio maybe we can simply removed the mount? Is an obsolete configuration?

**Comment by mdantonio** [comment]
If we want to move projects_configuration from rapydo-utils to rapydo-confs, we have to change:

- utilities/configuration.py -> read

- restapi/customization.py -> do_config

- utils/test/test_myyaml and utils/test/test_configuration

The very first question is: why does backend read configuration by it-self and not by using the utility function?

do_config shoud simply use read from utilities/configuration

I'm missing something? Or this is simply some old behaviour to be fixed?

**Comment by mdantonio** [comment]
At now projects_defaults is read by:

- utilities for testing purpose (to be removed)

- backend via restapi/customization.py -> do_config

- develop via utilities/configuration.py -> read

Since backed already read this file by it-self we could maintain do_config and simply replace the configuration mixing with the more advanced version in utilities/configuration.py -> mix

The projects_defaults could be mounted in backend container via compose:

```
confs/backend.yaml
    backend:
        volumes:
            ../submodules/rapydo-confs/projects_defaults.yaml:/code/projects_defaults.yml
```

This is a simple fix

But how to pass this file to develop? Develop requires this file to know GIT URIs of rapydo submodules

**Comment by pdonorio** [comment]
> So @pdonorio maybe we can simply removed the mount? Is an obsolete configuration?

Yep, it could easily be

> I'm missing something? Or this is simply some old behaviour to be fixed?

Idem

> But how to pass this file to develop? 

That's why it was packaged within `utilities`.
I have no idea how to do that without the package already installed. Pull it from github online?

**Comment by mdantonio** [comment]
Fixed the restclient paths

To back to projects_defaults:

At the moment dev is the only problem blocking the removal of core, because it is not coordinated by compose

Others (backend, celery and potentially frontend, that is not using projects_default but should) are coordinated by confs/backend.yaml and confs/frontend.yaml

Since these yaml files will be in rapydo-confs WITH projects_defaults, I think that it is quite safe to mount projects_defaults where needed (It is a configuration intra-repo ... that's good... what is bad are configurations inter-repos because they can create cyclic dependencies)

Just a consideration: dev is not part of the standard rapydo stack but is only intended for very few people who closely know the architecture. It's worth complicating the user architecture to simplify dev?
I think that we can easly duplicate in dev the few information extracted from projects_defaults. Dev simply requires the github path of subcomponents here:

https://github.com/rapydo/develop/blob/0.6.0/develop/cycles.py#L61-L66

```
from utilities.configuration import read
from develop import git
defaults = read()
repos = defaults.get('variables', {}).get('repos', {})
with path.cd(tools_version_path):
    git.clone(url=repos.get(toolname, {}).get('online_url', None))
```

Once hard-coded URIs in dev, how many times and how often we will change these URIs?

My proposal is:

1. hard-code subcomponents repo URL in dev
2. put a comment in projects_defaults variables.repo ("Ehi boy, if you change here, remember to fix in dev)
3. move projects_defaults in rapydo-confs
4. move core/confs in rapydo-confs
6. add rapydo-confs as new sub-module in vanilla/submodules
5. fix paths in confs/backend.yaml / frontend.yaml etc
6. mount projects_defaults in confs/backend.yaml
7. fix the controller to use yaml files from ./submodules/rapydo-confs/confs instead of ./confs
8. remove the read function in utilities.configuration
9. fix the do_config function in restapi (fix the new projects_default path and use the mix function in utilities.configuration)
10. remove from utils the unittests based on projects_defaults
11. rename rapydo-core as rapydo-tests
12. close this issue
13. work on issue #22 (Create command, as replacement of fork of core to create new projects)

**Comment by pdonorio** [comment]
> hard-code subcomponents repo URLs in dev

Since it's limited to me and you at the moment it could just be a parameter in the configuration file instead. Hard coded = really bad :)

Also we are actually considering the idea of dev as part of do, and that would definitely solve things right?

**Comment by mdantonio** [comment]
> Since it's limited to me and you at the moment it could just be a parameter in the configuration file instead. Hard coded = really bad :)

Ok ok, i modified the first point of the proposed todo list

> Also we are actually considering the idea of dev as part of do, and that would definitely solve things right?

I like this. But this solution does not solve all things because of a loop in required information (not strictly related to dev, but affecting the controller itself:

submodules urls are in projects_defaults.yaml

projects_defaults.yaml will be in rapydo-confs

rapydo-confs is a subcomponent

subcomponent are downloaded at init

init requires submodules urls


To break this loop the submodules urls (at least rapydo-confs) should be already available to controller prior to download subcomponents (i.e. hard-coded or configured in a controller-configuration)

Solution 1: the problem is fixed in controller, then dev uses controller to read such information

Solution 2: controller and dev are merged, then the problem is fixed once

**Comment by mdantonio** [comment]
I'm removing projects_default from utils.

Things to do in a near future:

- remove projects_default from files configured for package utilities
- mount projects_default in restclient

**Comment by mdantonio** [comment]
**Important**, vanilla side: you have to edit your compose files and change your paths.

Everything starting with "../" should be fixed, you can use variables:

`${SUBMODULE_DIR} `and `${VANILLA_DIR}`

**Comment by mdantonio** [comment]
Here the most is done, just waiting for last running unittests

What is missing:

- [ ] test in production mode
- [ ] read subcomponents repo URLs from dev configuration
- [ ] rename rapydo-core as rapydo-tests

**Comment by pdonorio** [comment]
Testing on eudat

**Comment by pdonorio** [comment]
Tested on eudat (in debug)

**Comment by mdantonio** [comment]
> Tested on eudat (in debug)

Do you tested in production too?

If everything works we can closed here!

**Comment by pdonorio** [comment]
I did
